packages/appsync/src/createAppSyncHooks.tsx

```diff
@@ -0,0 +1,198 @@
+import React, { useEffect, useState } from "react";
+import { createAuthLink } from "aws-appsync-auth-link";
+import { createSubscriptionHandshakeLink } from "aws-appsync-subscription-link";
+import {
+  gql,
+  ApolloClient,
+  InMemoryCache,
+  HttpLink,
+  ApolloLink,
+  useQuery as useQueryApollo,
+  useMutation as useMutationApollo,
+  useSubscription as useSubscriptionApollo,
+  QueryHookOptions,
+  MutationHookOptions,
+  ApolloProvider,
+  useApolloClient,
+  SubscriptionHookOptions,
+} from "@apollo/client";
+import { Files } from "./types";
+import { resolveFiles } from "./storageUtils";
+import { Auth } from "aws-amplify";
+
+export interface GraphqlProviderProps {
+  token?: string;
+}
+
+export const createAppSyncHooks = <
+  QT extends Record<string, { variables?: any; data: any }>,
+  MT extends Record<string, { variables?: any; data: any }>,
+  ST extends Record<string, { variables?: any; data: any }>
+>({
+  queries,
+  mutations,
+  subscriptions,
+  url,
+  region,
+  type,
+  refetchSubscriptions,
+}: {
+  queries: Record<keyof QT, string>;
+  mutations: Record<keyof MT, string>;
+  subscriptions: Record<keyof ST, string>;
+  url: string;
+  type: "AMAZON_COGNITO_USER_POOLS";
+  region: string;
+  refetchSubscriptions?: Partial<
+    Record<keyof ST, (keyof QT)[] | "all" | "active">
+  >;
+}) => {
+  const buildClient = ({ token }: { token?: string }) => {
+    const auth = {
+      type,
+      jwtToken: () => token || "",
+    };
+
+    const httpLink = ApolloLink.from([
+      createAuthLink({
+        url,
+        region,
+        auth,
+      }),
+      createSubscriptionHandshakeLink(
+        { url, region, auth },
+        new HttpLink({ uri: url })
+      ),
+    ]);
+
+    return new ApolloClient({
+      link: httpLink,
+      cache: new InMemoryCache(),
+    });
+  };
+
+  const RefetchSubscription: React.FC<{
+    mutation: keyof ST;
+    include: "all" | "active" | (keyof QT)[];
+  }> = ({ mutation, include }) => {
+    const client = useApolloClient();
+    useSubscription(mutation, {
+      onSubscriptionData: () => {
+        client.refetchQueries({
+          include: include as "all" | "active" | string[],
+        });
+      },
+    });
+    return null;
+  };
+
+  const RefetchSubscriptions: React.FC = () => {
```

Comment:
because hooks can't be inside a loop I had to create components that calls useSubscription but returns null instead. this is for the refetch functionality that listens to subscription events.